crates/astria-merkle/src/lib.rs

```diff
@@ -0,0 +1,718 @@
+//! A RFC 6962 compliant merkle tree with a flat representation.
+//!
+//! The intention
+//!
+//! The in-memory representation alternates between leaf nodes and branch nodes
+//! with a indexing shown below (the node with index 3 is the root).
+//! ```text
+//! 0
+//!   1
+//! 2
+//!    3
+//! 4
+//!   5
+//! 6
+//! ```
+//!
+//! # Indexing
+//! All indexing into the merkle tree rely on the following observation:
+//!
+//! A tree of depth 1 has the following addressing scheme:
+//! ```text
+//! 0       00
+//!   1     01
+//! 2       10
+//! ```
+//! And a tree of depth 2
+//! ```text
+//! 0      000
+//!   1    001
+//! 2      010
+//!    3   011
+//! 4      100
+//!   5    101
+//! 6      110
+//! ```
+//! And a subtree of depth 3
+//! ```text
+//! 0        0000
+//!    1     0001  <-- depth of subtree at index 1: 1
+//! 2        0010
+//!      3   0011  <-- depth of subtree at index 3: 2
+//! 4        0100
+//!    5     0101
+//! 6        0110
+//!        7 0111  <-- depth of subtree at root: 3
+//! 8        1000
+//!    9     1001
+//! 10       1010
+//!      11  1011  <-- depth of subtree at index 11: 2
+//! 12       1100
+//!    13    1101  <-- depth of subtree at index 13: 1
+//! 14       1110
+//! ```
+//! For the general case of a tree of depth `D+1` we then observe:
+//! ```text
+//! 011...1: root at index 2^D - 1
+//!  \____/
+//!    D number of ones
+//! \_____/
+//!   (D+1)-tree
+//!
+//! 0xxxxxx:
+//!  \____/
+//!    D-tree to the left of the root
+//!
+//! 1xxxxxx:
+//!  \____/
+//!    D-tree to the right of the root
+//! ```
+//! See [flat in order trees]( https://www.ietf.org/archive/id/draft-ietf-mls-protocol-14.html) for a full discussion and proof.
+//!
+//!
+//! # Further reading:
+//!
+//! + RFC 6962: <https://datatracker.ietf.org/doc/html/rfc6962>
+//! + RFC 7574 surpassing 6962: <https://datatracker.ietf.org/doc/rfc7574>
+//! + Hypercore whitepaper introducing flat binary trees: <https://www.datprotocol.com/deps/0002-hypercore/>
+//! + Traversing rachet trees: <https://www.ietf.org/archive/id/draft-ietf-mls-protocol-14.html>
+//! + Flat in-order trees (the blog post this crate is based on): <https://mmapped.blog/posts/22-flat-in-order-trees>
+
+#[cfg(test)]
+mod tests;
+
+use sha2::{
+    Digest as _,
+    Sha256,
+};
+
+/// Calculates `SHA256(0x00 | leaf)`
+#[must_use]
+pub fn hash_leaf(leaf: &[u8]) -> [u8; 32] {
+    let mut hasher = Sha256::new();
+    hasher.update([0x00_u8]);
+    hasher.update(leaf);
+    hasher.finalize().into()
+}
+
+/// Calculates `SHA256(0x01 || left || right)`.
+#[must_use]
+pub fn combine(left: &[u8], right: &[u8]) -> [u8; 32] {
+    let mut hasher = Sha256::new();
+    hasher.update([0x01_u8]);
+    hasher.update(left);
+    hasher.update(right);
+    hasher.finalize().into()
+}
+
+/// The proof that a node is included of a Merkle tree.
+///
+/// The proof is the concatenation of all merkle tree hashes from
+/// a node to the root.
```

Comment:
```suggestion
/// The proof is the concatenation of all sibling hashes required to
/// reconstruct the root from a given node.
```

something like this might be clearer?